### ISSUE TITLE
ULD wasserstein/topk_align feed student token IDs to a different teacher tokenizer

### ISSUE THREAD

**Issue by Amix29** [MakazhanAlpamys/Soup#681]
## Description

The non-aligned ULD strategies `wasserstein` and `topk_align` forward student `input_ids`
directly to the teacher. When the teacher vocabulary is smaller, Soup clamps the IDs into
range, but clamping does not translate tokens between vocabularies.

Affected revision: `ef2e02b51a1b34546823edce55715d28234805cb` (Soup 0.74.0).

Relevant code: [student IDs are copied and clamped before the teacher forward](https://github.com/MakazhanAlpamys/Soup/blob/ef2e02b51a1b34546823edce55715d28234805cb/src/soup_cli/trainer/distill.py#L606-L642).

Issue #258 added the separate `wasserstein_aligned` strategy and is complete. This report is
about the remaining non-aligned strategies: they still accept incompatible tokenizers instead
of restricting their contract or refusing the pair.

## Steps to reproduce

A real PyTorch CPU forward/backward was run with tiny local student and teacher models:

```text
student IDs for "hello cat dog":       [1, 6, 7]
IDs received by the teacher:           [1, 4, 4]
correct teacher IDs for the same text: [1, 2, 3]
training loss: finite (~1.86718)
backward: completes
```

The finite loss is part of the problem: the teacher can be conditioned on different or
corrupted text without a visible failure.

## Expected behavior

Each model should receive the same text encoded with its own tokenizer, or Soup should reject
a tokenizer pair that the selected strategy cannot handle safely.

## Impact

The student may distil a distribution produced from the wrong teacher context. Results from
cross-tokenizer ULD can therefore be semantically invalid even though training completes.

## Proposed fix and acceptance criteria

- [ ] Remove clamping as a semantic compatibility mechanism.
- [ ] Restrict `wasserstein` / `topk_align` to demonstrably ID-compatible tokenizers, or
      route incompatible pairs through a validated re-tokenization/alignment path.
- [ ] Fail early with a clear message when compatibility cannot be established.
- [ ] Add a tiny-model test with different token-ID maps that fails on the current path.
- [ ] Preserve the supported same-tokenizer fast path.

## Environment

Reproduced under Python 3.12.2 and PyTorch 2.12.1 CPU using the original `compute_loss` and ULD
kernel from the pinned checkout. No CUDA training session was required for the mechanism.

**Comment by MakazhanAlpamys** [comment]
**Confirmed, and it is lossier than you describe.**

The teacher tokenizer is never loaded for the non-aligned strategies — `distill.py:227` prints *"Loading tokenizer (student/teacher shared)"*, and `distill.py:362-372` loads a teacher tokenizer **only** when `uld_strategy == "wasserstein_aligned"`. The non-aligned branch reuses the student's `inputs` verbatim and applies `distill.py:625-630`:

```python
teacher_inputs["input_ids"] = teacher_inputs["input_ids"].clamp(max=int(_teacher_vocab) - 1)
```

Executed against the real `DistillTrainerWrapper`, two tiny local models with deliberately disjoint id maps (student vocab 12, teacher vocab 8), teacher `forward` instrumented:

```
uld_strategy = wasserstein
rendered text                 : 'ROLE user : hello cat EOT ROLE assistant : dog EOT '
student input_ids             : [3, 6, 4, 8, 9, 5, 3, 7, 4, 10, 5]
IDs the TEACHER actually saw  : [3, 6, 4, 7, 7, 5, 3, 7, 4, 7, 5]
teacher's OWN ids for the text: [5, 0, 6, 2, 3, 7, 5, 0, 6, 4, 7]
teacher DECODED what it saw   : 'cat : dog EOT EOT ROLE cat EOT dog EOT ROLE'
loss = 2.198  finite = True   backward OK
```

`topk_align` is identical.

## Two amendments, both against your framing

**The clamp is lossy, not merely wrong.** Student ids 8, 9 and 10 all collapse onto teacher id 7 — three distinct tokens become the same token, so the teacher distribution is not just misaligned, it is degenerate. And **when the teacher vocab is larger the clamp does nothing at all**; the ids are then simply wrong, with no collapse and no clue.

**This is the documented path, not an obscure one.** `docs/training.md:272-274`:

```
# Cross-tokenizer distillation — Llama -> Mistral, no shared vocab needed
soup train --config soup.yaml --uld-strategy wasserstein
```

So the feature is advertised for exactly the case where it silently misbehaves. The only console line is `Cross-tokenizer ULD enabled (strategy=wasserstein)`.

## Classification, since it matters for the fix

This is not the "validated then ignored" pattern — the flag is read and does change behaviour. It is the adjacent one: a **documented capability whose semantic contract is unimplemented**, with a crash-guard (`clamp`, added to stop an index error) standing in for the translation that was never written. `grep clamp tests/` finds nothing teacher- or vocab-related, so the clamp is not a defended contract either; it is scaffolding that became load-bearing.

There is a real question underneath this that I should answer rather than leave to a fixer: **is the ULD contract meant to be same-tokenizer-only?** No test or docstring says so, and `docs/training.md` says the opposite. My ruling: it is meant to work cross-tokenizer — that is the entire point of ULD — so this is a loss bug, not a docs bug. `wasserstein_aligned` exists and does load the teacher tokenizer, so the machinery is there; the non-aligned strategies need either to use it or to refuse the mismatch at setup with a message naming `wasserstein_aligned`.

Refusing loudly would be an acceptable interim fix and much smaller. Silently clamping is the one option that should not survive.

Labelled `bug` + `help wanted`, unclaimed.

Not settled: I used synthetic tokenizers with disjoint id maps to make the mismatch legible rather than downloading real Llama and Mistral checkpoints. The mechanism is a `clamp` on ids and is tokenizer-independent, but nobody has run it on the model pair the docs name.

**Comment by MakazhanAlpamys** [comment]
Fixed in `a1eab0f` by @AmirF194 in #704.

`wasserstein` and `topk_align` now refuse a tokenizer pair that is not interchangeable, at `setup()`, naming `wasserstein_aligned` as the working alternative.

**Your report was understated and I said so when confirming it**: the clamp does not translate tokens between vocabularies, so a mismatched pair trained on logits the teacher computed for *different text* while reporting a finite, plausible loss — and the clamp did nothing at all when the teacher vocabulary was the larger one. Silently wrong, not a crash.

Four mutations killed, including the one that matters most here: **unwiring the call from `setup()` while leaving the helper intact** fails 2 tests. That is the standing guard against this repo's most persistent defect — a correct helper the command never calls (#363, #423, #659, #672, #675).

The clamp removal is complete rather than partial: `same_tokenizer` requires equal `vocab_size` **and** identical ids over `PROBE_CORPUS`, so the index error it guarded cannot occur once the check passes, and `_teacher_vocab` had no other reader.

**Note that this narrows two shipped, documented strategies to same-tokenizer only.** No bundled recipe, template or example sets `uld_strategy`, so nothing shipped breaks — but a hand-written config that appeared to work will now refuse. `docs/training.md` moves its Llama -> Mistral example onto `wasserstein_aligned`, and `docs/commands.md:313`, which still called these two "cross-tokenizer" and would have walked a reader straight into the new error, was corrected on review.

Reported by @Amix29.
